rcgen/src/key_pair.rs

```diff
@@ -319,6 +323,167 @@ impl KeyPair {
+	/// However, sometimes multiple signature algorithms fit for the same DER key. In those instances,
+	/// you can use this function to precisely specify the `SignatureAlgorithm`.
+	///
+	/// You can use [`rustls_pemfile::private_key`] to get the `key` input. If
+	/// you have already a byte slice, just calling `try_into()` will convert it to a [`PrivateKeyDer`].
+	///
+	/// [`rustls_pemfile::private_key`]: https://docs.rs/rustls-pemfile/latest/rustls_pemfile/fn.private_key.html
+	#[cfg(feature = "crypto")]
+	pub fn from_der_and_sign_algo(
+		key: &PrivateKeyDer<'_>,
+		alg: &'static SignatureAlgorithm,
+	) -> Result<Self, Error> {
+		#[cfg(all(feature = "ring", not(feature = "aws_lc_rs")))]
+		{
+			if let PrivateKeyDer::Pkcs8(key) = key {
+				Self::from_pkcs8_der_and_sign_algo(key, alg)
+			} else {
+				Err(Error::CouldNotParseKeyPair)
+			}
+		}
+		#[cfg(feature = "aws_lc_rs")]
+		{
+			let is_pkcs8 = matches!(key, PrivateKeyDer::Pkcs8(_));
+
+			let rsa_key_pair_from = if is_pkcs8 {
+				RsaKeyPair::from_pkcs8
+			} else {
+				RsaKeyPair::from_der
+			};
+
+			let serialized_der = key.secret_der().to_vec();
+
+			let kind = if alg == &PKCS_ED25519 {
+				KeyPairKind::Ed(Ed25519KeyPair::from_pkcs8_maybe_unchecked(&serialized_der)._err()?)
+			} else if alg == &PKCS_ECDSA_P256_SHA256 {
+				KeyPairKind::Ec(ecdsa_from_private_key_der(
+					&signature::ECDSA_P256_SHA256_ASN1_SIGNING,
+					&serialized_der,
+				)?)
+			} else if alg == &PKCS_ECDSA_P384_SHA384 {
+				KeyPairKind::Ec(ecdsa_from_private_key_der(
+					&signature::ECDSA_P384_SHA384_ASN1_SIGNING,
+					&serialized_der,
+				)?)
+			} else if alg == &PKCS_ECDSA_P521_SHA512 {
+				KeyPairKind::Ec(ecdsa_from_private_key_der(
+					&signature::ECDSA_P521_SHA512_ASN1_SIGNING,
+					&serialized_der,
+				)?)
+			} else if alg == &PKCS_RSA_SHA256 {
+				let rsakp = rsa_key_pair_from(&serialized_der)._err()?;
+				KeyPairKind::Rsa(rsakp, &signature::RSA_PKCS1_SHA256)
+			} else if alg == &PKCS_RSA_SHA384 {
+				let rsakp = rsa_key_pair_from(&serialized_der)._err()?;
+				KeyPairKind::Rsa(rsakp, &signature::RSA_PKCS1_SHA384)
+			} else if alg == &PKCS_RSA_SHA512 {
+				let rsakp = rsa_key_pair_from(&serialized_der)._err()?;
+				KeyPairKind::Rsa(rsakp, &signature::RSA_PKCS1_SHA512)
+			} else if alg == &PKCS_RSA_PSS_SHA256 {
+				let rsakp = rsa_key_pair_from(&serialized_der)._err()?;
+				KeyPairKind::Rsa(rsakp, &signature::RSA_PSS_SHA256)
+			} else {
+				panic!("Unknown SignatureAlgorithm specified!");
+			};
+
+			Ok(KeyPair {
+				kind,
+				alg,
+				serialized_der,
+			})
+		}
+	}
+
+	/// Parses the key pair from the DER format
+	///
+	/// You can use [`rustls_pemfile::private_key`] to get the `key` input. If
+	/// you have already a byte slice, just calling `try_into()` will convert it to a [`PrivateKeyDer`].
+	///
+	/// [`rustls_pemfile::private_key`]: https://docs.rs/rustls-pemfile/latest/rustls_pemfile/fn.private_key.html
+	#[cfg(feature = "crypto")]
+	pub(crate) fn from_raw(
```

Comment:
One suggestion: if possible I would move this code into a `impl TryFrom<&PrivateKeyDer> for KeyPair`, skipping the `from_raw()` indirection.

Reply:
I mean pretty much moving all of it, so removing `from_raw()` (which we definitely don't want to duplicate). Is that not feasible?